engine/src/main/scala/cromwell/engine/io/nio/NioFlow.scala

```diff
@@ -105,30 +107,60 @@ class NioFlow(parallelism: Int,
     ()
   }
 
-  private def readAsString(read: IoContentAsStringCommand) = IO {
-    new String(
-      read.file.limitFileContent(read.options.maxBytes, read.options.failOnOverflow),
-      StandardCharsets.UTF_8
-    )
-      .replaceAll("\\r\\n", "\\\n")
+  private def readAsString(read: IoContentAsStringCommand): IO[String] = {
+    def checkHash(value: String, fileHash: FileHash): IO[ChecksumResult] = {
+      // Disable checksum validation if failOnOverflow is false. We might not read
+      // the entire stream, in which case the checksum will definitely fail.
+      // Ideally, we would only disable checksum validation if there was an actual
+      // overflow, but we don't know that here.
+      if (!read.options.failOnOverflow) return IO.pure(ChecksumSuccess())
+
+      val hash = fileHash.hashType.calculateHash(value)
+      if (hash == fileHash.hash) IO.pure(ChecksumSuccess())
+      else IO.pure(ChecksumFailure(hash))
+    }
+
+    def readFile: IO[String] = {
+      for {
+        fileHash <- getHash(read.file)
+        uncheckedValue <- IO {
+          new String(
+            read.file.limitFileContent(read.options.maxBytes, read.options.failOnOverflow),
+            StandardCharsets.UTF_8
+          )
+        }
+        checksumResult <- checkHash(uncheckedValue, fileHash)
+        verifiedValue <- checksumResult match {
+          case _: ChecksumSuccess => IO.pure(uncheckedValue)
+          case failure: ChecksumFailure => IO.raiseError(
+            ChecksumFailedException(
+              s"Failed checksum for: ${read.file}. expected: '$fileHash' actual: '${failure.calculatedHash}'"))
+        }
+      } yield verifiedValue
+    }
+
+    readFile.map(_.replaceAll("\\r\\n", "\\\n"))
```

Comment:
It looks like this will checksum-validate all files we read using this method, not just DRS files. I think we planned to do DRS files only, but I'm certainly willing to think about whether it makes sense to do everything. I worry about out-of-date hashes in S3, etc. causing problems for us.

Reply:
Ah, you're right. I knew we had agreed to not do checksums on just signed URLs, but forgot that there was a midpoint between that and checksumming everything... and that is to checksum all DRS access, whether that results in signed URL or direct GCS bucket access. Changing course on this will greatly help work around tests that are currently failing (like OSS object access).